vault/acl.go

```diff
@@ -47,25 +49,88 @@ func NewACL(policies []*Policy) (*ACL, error) {
 			// Check for an existing policy
 			raw, ok := tree.Get(pc.Prefix)
 			if !ok {
-				tree.Insert(pc.Prefix, pc.CapabilitiesBitmap)
+				tree.Insert(pc.Prefix, pc.Permissions)
 				continue
 			}
-			existing := raw.(uint32)
+
+			// these are the ones already in the tree
+			existingPerms := raw.(*Permissions)
 
 			switch {
-			case existing&DenyCapabilityInt > 0:
+			case existingPerms.CapabilitiesBitmap&DenyCapabilityInt > 0:
 				// If we are explicitly denied in the existing capability set,
 				// don't save anything else
+				continue
 
-			case pc.CapabilitiesBitmap&DenyCapabilityInt > 0:
+			case pc.Permissions.CapabilitiesBitmap&DenyCapabilityInt > 0:
 				// If this new policy explicitly denies, only save the deny value
-				tree.Insert(pc.Prefix, DenyCapabilityInt)
+				pc.Permissions.CapabilitiesBitmap = DenyCapabilityInt
+				pc.Permissions.AllowedParameters = nil
+				pc.Permissions.DeniedParameters = nil
+				goto INSERT
 
 			default:
 				// Insert the capabilities in this new policy into the existing
 				// value
-				tree.Insert(pc.Prefix, existing|pc.CapabilitiesBitmap)
+				pc.Permissions.CapabilitiesBitmap = existingPerms.CapabilitiesBitmap | pc.Permissions.CapabilitiesBitmap
+			}
+
+			if len(existingPerms.AllowedParameters) > 0 {
```

Comment:
I think we should check parameters case-insensitively. Although technically this could maybe lead to undesired behavior in the `generic` backend, it also ensures we're protected in the case of such a change to framework, or to any backend looking at the raw request parameters in a case-insensitive fashion.

I put this comment here instead of below because even in this case, then, we need to ensure that if pc.Permisisons.AllowedParameters is not nil that we create a lowercase version.

It's also possible that the easiest way to do this is to simply ensure when we parse the HCL that the map that goes into the Permissions struct isn't merely a copy but instead `strings.ToLower`s as it goes. Then we just need to ensure that in the AllowOperation calls we are also checking this.